File: libs/edis-ts-contracts/src/governance.ts
/**
 * Zod mirror of `edis_contracts/governance.py` (audit + explainability),
 * `LineageEvent` from `edis_contracts/events.py`, and the authz boundary types
 * from `edis_contracts/security.py`.
 *
 * `LineageEvent` lives here (not in canonical) because it is the payload of
 * `edis.governance.lineage.v1`. The security types are the shapes the BFF derives
 * from a verified JWT and attaches to every request.
 */
import { z } from "zod";
import { anyMap, datetime, schemaVersion, uuid } from "./common.js";

/** Mirror of `LineageEvent` -- payload of `edis.governance.lineage.v1`. */
export const LineageEventSchema = z.object({
  lineage_id: uuid(),
  tenant_id: z.string(),
  run_id: uuid(),
  inputs: z.array(z.record(z.string(), z.unknown())).default([]),
  outputs: z.array(z.record(z.string(), z.unknown())).default([]),
  stage: z.string(),
  occurred_at: datetime(),
  schema_version: schemaVersion(),
});
export type LineageEvent = z.infer<typeof LineageEventSchema>;

/** Mirror of `AuditEvent` -- payload of `edis.governance.audit.v1`. */
export const AuditEventSchema = z.object({
  audit_id: uuid(),
  occurred_at: datetime(),
  tenant_id: z.string(),
  actor: anyMap().default({}),
  action: z.enum([
    "DATA_READ",
    "DATA_WRITE",
    "AI_DECISION",
    "AI_QUERY",
    "AUTH_DENY",
    "RBAC_CHANGE",
    "EXPORT",
  ]),
  resource: anyMap().default({}),
  outcome: z.enum(["ALLOW", "DENY", "ERROR"]),
  reason: z.string().nullish(),
  decision_id: uuid().nullish(),
  trace_id: z.string().nullish(),
  schema_version: schemaVersion(),
});
export type AuditEvent = z.infer<typeof AuditEventSchema>;

/** Mirror of `Evidence` (immutable value snapshot + live pointer). */
export const EvidenceSchema = z.object({
  evidence_id: uuid(),
  kind: z.string(),
  summary: z.string(),
  snapshot: anyMap().default({}),
  ref: z.record(z.string(), z.unknown()).nullish(),
  schema_version: schemaVersion(),
});
export type Evidence = z.infer<typeof EvidenceSchema>;

/** Mirror of `Decision` -- explainability record linking an AI decision to evidence. */
export const DecisionSchema = z.object({
  decision_id: uuid(),
  tenant_id: z.string(),
  decision_type: z.enum([
    "finding_narrative",
    "recommendation",
    "copilot_answer",
  ]),
  subject_id: uuid(),
  actor: anyMap().default({}),
  rationale: z.string(),
  evidence: z.array(EvidenceSchema).default([]),
  created_at: datetime(),
  schema_version: schemaVersion(),
});
export type Decision = z.infer<typeof DecisionSchema>;

// --- security.py: authz boundary types ---

/** Mirror of `Role = Literal[...]`. */
export const RoleSchema = z.enum([
  "viewer",
  "analyst",
  "operator",
  "auditor",
  "admin",
]);
export type Role = z.infer<typeof RoleSchema>;

/** Mirror of `Actor`. */
export const ActorSchema = z.object({
  type: z.enum(["user", "service", "system", "copilot"]),
  id: z.string(),
  roles: z.array(z.string()).default([]),
});
export type Actor = z.infer<typeof ActorSchema>;

/** Mirror of `ResourceRef`. */
export const ResourceRefSchema = z.object({
  type: z.string(),
  id: z.string().nullish(),
  columns: z.array(z.string()).nullish(),
});
export type ResourceRef = z.infer<typeof ResourceRefSchema>;

/** Mirror of `SecurityContext` -- the authenticated principal from the JWT. */
export const SecurityContextSchema = z.object({
  tenant_id: z.string(),
  user_id: z.string(),
  roles: z.array(z.string()).default([]),
  scopes: z.array(z.string()).default([]),
  token_id: z.string().nullish(),
});
export type SecurityContext = z.infer<typeof SecurityContextSchema>;

File: libs/edis-contracts/edis_contracts/governance.py
"""Governance contracts: audit events and the explainability store.

Every data access and AI decision emits an :class:`AuditEvent`. Every AI decision
links to a :class:`Decision` + :class:`Evidence` record whose snapshots are
immutable -- so a narrative is reproducible even if the source data later mutates.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class AuditEvent(BaseModel):
    """Payload of ``edis.governance.audit.v1`` -- emitted by every layer.

    ``audit_id`` is the idempotency key (the audit consumer dedupes on it).
    """

    audit_id: UUID
    occurred_at: datetime
    tenant_id: str
    actor: dict = Field(default_factory=dict)  # {type, id, roles}
    action: Literal[
        "DATA_READ",
        "DATA_WRITE",
        "AI_DECISION",
        "AI_QUERY",
        "AUTH_DENY",
        "RBAC_CHANGE",
        "EXPORT",
    ]
    resource: dict = Field(default_factory=dict)  # {type, id, columns?}
    outcome: Literal["ALLOW", "DENY", "ERROR"]
    reason: str | None = None
    decision_id: UUID | None = None  # link to explainability for AI_* actions
    trace_id: str | None = None
    schema_version: int = 1


class Evidence(BaseModel):
    """An immutable value snapshot plus a live pointer back to the source."""

    evidence_id: UUID
    kind: str  # "metric" | "finding" | "recommendation" | "tool_result"
    summary: str
    snapshot: dict = Field(default_factory=dict)  # frozen values (reproducible)
    ref: dict | None = None  # live pointer
    schema_version: int = 1


class Decision(BaseModel):
    """An explainability record linking an AI decision to its evidence trail."""

    decision_id: UUID
    tenant_id: str
    decision_type: Literal["finding_narrative", "recommendation", "copilot_answer"]
    subject_id: UUID  # finding_id / recommendation_id / answer_id
    actor: dict = Field(default_factory=dict)
    rationale: str
    evidence: list[Evidence] = Field(default_factory=list)
    created_at: datetime
    schema_version: int = 1
